R/SandwichLayer.R

```diff
@@ -156,26 +199,77 @@ setMethod("[", "PreSandwichLayer",
     warning("prediction from a rank-deficient fit may be misleading")
     X <- X[, piv, drop = FALSE]
   }
-
+  
   # TODO: support predict(..., type = "response"/"link"/other?)
   xb <- drop(X %*% model$coefficients[piv])
 
-  # this branch applies to (at least) `glm`, `survey::surveyglm`,
-  # `robustbase::glmrob` and `gam` models
-  if (inherits(model, "glm")) {
-    ca <- model$family$linkinv(xb)
-    pred_gradient <- model$family$mu.eta(xb) * X
-  } else if (inherits(model, "lm") | inherits(model, "lmrob")) {
-    # `lm` doesn't have a `family` object, but we know its prediction gradient
-    ca <- xb
-    pred_gradient <- X
-  } else {
-    stop("`model` must inherit from a `glm`, `lm`, or `lmrob` object")
-  }
+  return(new("PreSandwichLayer",
+             model$family$linkinv(xb),
+             fitted_covariance_model = model,
+             prediction_gradient = model$family$mu.eta(xb) * X))
+}
+.S3method(".make_PreSandwichLayer", "glm", .make_PreSandwichLayer.glm)
 
-  return(list("ca" = ca,
-              "prediction_gradient" = pred_gradient))
+##' @keywords internal
+.make_PreSandwichLayer.glmrob <- function(model, newdata = NULL, ...) {
+  # if (is.null(newdata)) {
+  #   newdata <- stats::model.frame(model)
+  # }
+  # 
+  # if (!is.data.frame(newdata)) {
+  #   stop("If supplied, `newdata` must be a dataframe")
+  # }
+  # 
+  model_terms <- tryCatch(
+    terms(model),
+    error = function(e) stop("`model` must have `terms` method", call. = FALSE)
+  )
+  # newdata <- tryCatch({
+  #   try_terms <- terms(newdata) # terms will be the same if newdata generated with model's model.frame
+  #   if (!is.null(try_terms)) newdata
+  #   }, error = function(e) {
+  #     stats::model.frame(stats::delete.response(model_terms),
+  #                        data = newdata,
+  #                        na.action = na.pass,
+  #                        xlev = model$xlevels) # xlev arg follows `predict.lm`
+  # })
+  
+  # X <- stats::model.matrix(stats::delete.response(model_terms),
+  #                          data = newdata,
+  #                          contrasts.arg = model$contrasts) # contrasts.arg also follows `predict.lm`
+  
+  if (is.null(newdata)) newdata <- .get_data_from_model("cov_adj", formula(model))
+  
+  mf <- stats::model.frame(stats::delete.response(terms(model)),
+                           data = newdata,
+                           na.action = na.pass,
+                           xlev = model$xlevels,
+                           ...)
+
+  X <- stats::model.matrix(stats::delete.response(terms(model)),
+                           data = mf,
+                           contrasts.arg = model$contrasts, # contrasts.arg also follows `predict.lm`
+                           ...)
+  
+  # use the `stats` package's method for handling model fits not of full rank
+  QR <- qr(stats::model.matrix(model))
```

Comment:
Hadn't communicated this previously, but it occurs to me now that the rows of this model matrix should be weighted by "robustness weights", i.e. `model$w.r`, prior to computation of the QR.  That brings us closer to what happens in the final iteration of IRWLS.